File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        pal_start = pal_max_len = 0
        n = len(s)

        for i in range(n):
            pal_start1, pal_len1 =  self.expand_around_centers(s, i, i)
            pal_start2, pal_len2 = self.expand_around_centers(s, i, i + 1)

            if pal_max_len < pal_len1:
                pal_start, pal_max_len = pal_start1, pal_len1
            if pal_max_len < pal_len2:
                pal_start, pal_max_len = pal_start2, pal_len2

        return s[pal_start:pal_start + pal_max_len]

    def expand_around_centers(self, s, l, r):
        while (l >= 0 and r < len(s) and s[l] == s[r]):
            l -= 1
            r += 1

        return l + 1, r - l - 1

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.rb
# @param {String} s
# @return {String}
def longest_palindrome(s)
  n = s.length
  pal_start = 0
  pal_max_len = 0

  (0...n).each do |i|
    start1, len1 = expand_around_centers(s, i, i)       # Odd-length
    start2, len2 = expand_around_centers(s, i, i + 1)   # Even-length

    if len1 > pal_max_len
      pal_start = start1
      pal_max_len = len1
    end

    if len2 > pal_max_len
      pal_start = start2
      pal_max_len = len2
    end
  end

  s[pal_start, pal_max_len]
end

def expand_around_centers(s, left, right)
  while left >= 0 && right < s.length && s[left] == s[right]
    left -= 1
    right += 1
  end

  return left + 1, right - left - 1
end
